nexus/db-queries/src/db/datastore/db_metadata.rs

```diff
@@ -213,93 +215,157 @@ impl DataStore {
-                "target_version" => target_version.to_string(),
+                "current_version" => %current_version,
+                "target_version" => %target_version,
             );
 
             let target_dir = config.schema_dir.join(target_version.to_string());
 
-            let schema_change =
-                all_sql_for_version_migration(&target_dir).await?;
+            let schema_change = all_sql_for_version_migration(&target_dir)
+                .await
+                .map_err(anyhow::Error::msg)
+                .context("Failed to access all SQL statements for a version migration")?;
 
-            // Confirm the current version, set the "target_version"
-            // column to indicate that a schema update is in-progress.
+            // Iterate over each individual file that comprises a schema change.
             //
-            // Sets the following:
-            // - db_metadata.target_version = new version
-            self.prepare_schema_update(&current_version, &target_version)
+            // For example, this contains "up01.sql", "up02.sql", etc.
+            //
+            // While this update happens, the "current_version" will remain
+            // the same (unless another Nexus concurrently completes the
+            // update), but the "target_version" will keep shifting on each
+            // incremental step.
+            let mut prior_step_version = None;
+            for (i, SchemaUpgradeStep { path: _, sql }) in
+                schema_change.steps.iter().enumerate()
+            {
+                // Durably store our progress in the target version as the
+                // pre-release extension of "target_version".
+                target_version.0.pre =
+                    semver::Prerelease::new(&format!("step.{i}"))
+                        .context("Cannot parse step as semver pre-release")?;
+
+                info!(
+                    log,
+                    "Considering found target version";
+                    "observed_target_version" => ?found_target_version,
+                    "attempted_target_version" => %target_version,
+                );
+
+                if let Some(found_target_version) =
+                    found_target_version.as_ref()
+                {
+                    // This case only occurs if an upgrade failed and needed to
+                    // restart, for whatever reason. We skip all the incremental
+                    // steps that we know have completed.
+                    if found_target_version > &target_version {
+                        warn!(
+                            log,
+                            "Observed target version greater than this upgrade step. Skipping.";
+                            "observed_target_version" => %found_target_version,
+                            "attempted_target_version" => %target_version,
+                        );
+                        continue;
```

Comment:
Do we need to set `prior_step_version` before `continue`ing? If so, could we rework the tests to fail without doing so? (That might be easier after #5274, when we can pass in the schema versions instead of a directory to read.) This sequence seems like it won't work:

* Update to `N.0.0-step.1` is successful
* Update to `N.0.0-step.2` fails; we bail out
* Next attempt, we land here and note that `N.0.0-step.2 > N.0.0-step.1`, so we `continue`; `prior_step_version` is still `None`
* We move on to `N.0.0-step.2`, and call `prepare_schema_update`. Because `prior_step_version` is `None`, its `valid_prior_targets` list only includes `N.0.0`, not `N.0.0-step.1`, so it fails.

Reply:
Okay, keep me honest here - I'm on board with a way to make this simpler, but:

I think it's okay to keep the `prior_step_version` as "the latest version we think **our Nexus** has applied".

@jgallagher , in the case you mention, once we get to that fourth step (calling `prepare_schema_update`) , you're right that our `prior_step_version` is `None`, but the `to_version` parameter would still be `N.0.0-step.2` (we're trying to incrementally work on that step) so it would be included in `valid_prior_targets`. That would also be the target version stored in the DB, since that's the last thing we updated before failing on the prior attempt to update the schema.

(I think this was true before my merge as well, but less obvious, because I was mutating `target_version`. Now, I'm using a different variable named `target_step_version` to make this a little more clear).

Reply:
Also, the test I added below really should be catching cases like this (and has, locally, when I've gotten it wrong) because it does trigger this "Fail a step and force retry" behavior.

Reply:
> `to_version` parameter would still be `N.0.0-step.2`

Is this guaranteed? What if we succeed in updating to `N.0.0-step.1`, and then `prepare_schema_update` fails to set the new target version of `N.0.0-step.2`?

Reply:
If that's the case, we'd re-apply "step 1", redoing the work of `N.0.0-step.1`, which should be fine since individual steps must be idempotent anyway, right?

Also, regardless: I'm taking this feedback, and refactored a bit of this to try to make it easier to follow -- though I think the functionality is the same.

Reply:
Yes, you're right. Thanks for sticking with me on this one, and for the refactor - I find it more clearly correct for sure.